File: cortex/tests/fixtures/ros2_workspace/src/demo_pkg/demo_pkg/talker.py
# Stub rclpy source used only by kg tests; not executed.
#
# Exercises the surface area of the ``ros2_topology`` (Python half)
# extractor, deliberately mirroring ``src/talker.cpp`` so the shared
# contract tests line up:
#
# - ``rclpy.node.Node`` subclass whose ``__init__`` passes the runtime
#   name to ``super().__init__("talker")``
# - ``self.create_publisher(std_msgs.msg.String, "chatter", 10)``
# - ``self.create_service(demo_pkg.srv.Ping, "ping", handler)``
# - ``self.create_client(demo_pkg.srv.Ping, "ping_client")``
# - ``ActionServer(self, demo_pkg.action.Fibonacci, "fibonacci", handler)``
# - ``self.declare_parameter("period_ms", 500)``
# - ``self.get_parameter("undeclared_flag")`` (declared: False path)
# - a dynamic (non-literal) topic name for the fallback id path

import rclpy
from rclpy.action import ActionServer
from rclpy.node import Node

import demo_pkg.action
import demo_pkg.srv
import std_msgs.msg

class Talker(Node):
    def __init__(self) -> None:
        super().__init__("talker")
        self._topic_name = "chatter_dyn"
        self.publisher_ = self.create_publisher(
            std_msgs.msg.String, "chatter", 10
        )
        self.service_ = self.create_service(
            demo_pkg.srv.Ping, "ping", self._handle_ping
        )
        self.client_ = self.create_client(
            demo_pkg.srv.Ping, "ping_client"
        )
        self.action_server_ = ActionServer(
            self,
            demo_pkg.action.Fibonacci,
            "fibonacci",
            self._handle_fibonacci,
        )
        self.declare_parameter("period_ms", 500)
        self.get_parameter("undeclared_flag")
        # Dynamic topic name: non-literal second argument.
        self.dyn_publisher_ = self.create_publisher(
            std_msgs.msg.String, self._topic_name, 10
        )

    def _handle_ping(self, request, response):
        return response

    def _handle_fibonacci(self, goal_handle):
        return goal_handle

def main() -> None:
    rclpy.init()
    node = Talker()
    rclpy.spin(node)

if __name__ == "__main__":
    main()

File: cortex/tests/fixtures/ros2_workspace/src/demo_pkg/src/talker.cpp
// Stub ROS2 C++ source used only by kg tests; not built.
//
// Exercises the surface area of the ``ros2_topology`` (C++ half) extractor:
//   - rclcpp::Node subclass with a runtime name passed to the Node()
//     super-constructor
//   - create_publisher<std_msgs::msg::String>("chatter", ...)
//   - create_service<demo_pkg::srv::Ping>("ping", ...)
//   - create_client<demo_pkg::srv::Ping>("ping_client")
//   - rclcpp_action::create_server<demo_pkg::action::Fibonacci>(...)
//   - declare_parameter<int>("period_ms", 500)
//   - get_parameter("undeclared_flag") (declared: false path)
//   - a dynamic (non-literal) topic name for the fallback id path
//   - RCLCPP_COMPONENTS_REGISTER_NODE(demo_pkg::Talker) at file scope

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "std_msgs/msg/string.hpp"
#include "demo_pkg/srv/ping.hpp"
#include "demo_pkg/action/fibonacci.hpp"

namespace demo_pkg {

class Talker : public rclcpp::Node {
 public:
  Talker() : Node("talker") {
    publisher_ = create_publisher<std_msgs::msg::String>("chatter", 10);
    service_ = create_service<demo_pkg::srv::Ping>(
        "ping", [](const auto, auto) {});
    client_ = create_client<demo_pkg::srv::Ping>("ping_client");
    action_server_ = rclcpp_action::create_server<demo_pkg::action::Fibonacci>(
        this, "fibonacci", nullptr, nullptr, nullptr);
    declare_parameter<int>("period_ms", 500);
    auto flag = get_parameter("undeclared_flag");
    // Dynamic topic name: non-literal first argument.
    dyn_publisher_ = create_publisher<std_msgs::msg::String>(
        topic_name_, 10);
  }

 private:
  std::string topic_name_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr dyn_publisher_;
  rclcpp::Service<demo_pkg::srv::Ping>::SharedPtr service_;
  rclcpp::Client<demo_pkg::srv::Ping>::SharedPtr client_;
  rclcpp_action::Server<demo_pkg::action::Fibonacci>::SharedPtr action_server_;
};

}  // namespace demo_pkg

RCLCPP_COMPONENTS_REGISTER_NODE(demo_pkg::Talker)
